hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,715 @@
+	{
+		LastValues[index] = value;
+		return true;
+	}
+	return false;
+}
+
+//option management
+void getOption(TOptionMap& Option, const char* optionName, int& value)
+{
+	value = std::stoi(Option[optionName].c_str());
+}
+void getOption(TOptionMap& Option, const char* optionName, long& value)
+{
+	value = std::stol(Option[optionName].c_str());
+}
+void getOption(TOptionMap& Option, const char* optionName, double& value)
+{
+	value = std::stof(Option[optionName].c_str());
+}
+void getOption(TOptionMap& Option, const char* optionName, std::string& value)
+{
+	value = (Option[optionName]);
+}
+
+void setOption(TOptionMap& Option, const char* optionName, int  value)
+{
+	Option[optionName] = std::to_string(value);
+}
+void setOption(TOptionMap& Option, const char* optionName, double  value)
+{
+	Option[optionName] = std_format("%4.1f", value);;
+}
+void setOption(TOptionMap& Option, const char* optionName, const char* value)
+{
+	Option[optionName] = value;
+}
+void setOption(TOptionMap& Option, const char* optionName, std::string& value)
+{
+	Option[optionName] = value;
+}
+
+
+VirtualThermostat* m_VirtualThermostat;
+
+VirtualThermostat::VirtualThermostat(const int ID)
+{
+	m_HwdID = ID;
+	m_VirtualThermostat = this;
+	//thermostat mode string
+	SetAvailableMode(AVALAIBLE_MODE);
+
+}
+
+VirtualThermostat::~VirtualThermostat()
+{
+	m_bIsStarted = false;
+}
+bool VirtualThermostat::StartHardware()
+{
+	RequestStart();
+
+	m_stoprequested = false;
+
+	//Start worker thread
+	//Start worker thread
+	m_thread = std::make_shared<std::thread>(&VirtualThermostat::Do_Work, this);
+	SetThreadNameInt(m_thread->native_handle());
+
+	sOnConnected(this);
+	m_bIsStarted = true;
+	return (m_thread != NULL);
+}
+bool VirtualThermostat::StopHardware()
+{
+	m_stoprequested = true;
+	{
+		RequestStop();
+		m_thread->join();
+		m_thread.reset();
+	}
+	m_bIsStarted = false;
+	return true;
+}
+void VirtualThermostat::Do_Work()
+{
+	int sec_counter = 0;
+	Log(LOG_STATUS, "VirtualThermostat: Worker started...");
+	while (IsStopRequested(1000) == false)
+	{
+		if (m_stoprequested)
+			break;
+		sec_counter++;
+
+		if (sec_counter % 12 == 0) {
+			m_LastHeartbeat = mytime(NULL);
+		}
+		time_t atime = mytime(NULL);
+		struct tm ltime;
+		localtime_r(&atime, &ltime);
+#if 1
+		if (ltime.tm_min != m_ScheduleLastMinute)
+		{
+			ScheduleThermostat(ltime.tm_min);
+			m_ScheduleLastMinute = ltime.tm_min;
+		}
+#else
+		//test : call every 6 secconds
+		if (ltime.tm_sec % 6 == 0)
+		{
+			ScheduleThermostat(ltime.tm_sec / 6);
+			m_ScheduleLastMinute = ltime.tm_sec;
+		}
+#endif
+	}
+	Log(LOG_STATUS, "VirtualThermostat: Worker stopped...");
+}
+bool VirtualThermostat::WriteToHardware(const char* pdata, const unsigned char length)
+{
+	return true;
+}
+
+//time for the power time modulation in minute
+#define MODULATION_DURATION 10
+//  number of step in percent %
+#define MODULATION_STEP     10
+
+//guve the output power swicth value in the time from the modulation percent
```

Comment:
```suggestion
//give the output power switch value in the time from the modulation percent
```
Are there boolean values otherwise?

Reply:
ok.
Yes